### PR TITLE
Khaled DB W3

### PR DESCRIPTION
**Pull request by KhaledAlsaleh**

### CODE REVIEW

File: Week3/homework/ex1-Normalization/1-violate1NF.txt [hunk 1]
```diff
@@ -0,0 +1,15 @@
+What columns violate 1NF?
+
+1NF (4 rules):
+
+    1- Single valued columns (each column should have atomic value, no multiple values):
+    These columns (member_address,food_code,food_description) violate this rule (they have multiple values).
```
Comment (mogaleaf) [inline Week3/homework/ex1-Normalization/1-violate1NF.txt:6]: I agree for food_code and food_description but the member_address could be considered as correct as it represents only ONE address.

File: Week3/homework/ex1-Normalization/1-violate1NF.txt [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+    2- Column domain (for any column) should not change:
```
Comment (mogaleaf) [inline Week3/homework/ex1-Normalization/1-violate1NF.txt:7]: yes soo bad

File: Week3/homework/ex1-Normalization/2-entitiesExtracted.txt [hunk 3]
```diff
@@ -0,0 +1,9 @@
+To make the table follow 1NF rules, I have to extracting these entities:
+    (member_address,food_code,food_description,dinner_date)
+
+1- member_address: Extract this column into 2 column (building_no hold integer data type, building_name hold varchar data type).
```
Comment (mogaleaf) [inline Week3/homework/ex1-Normalization/2-entitiesExtracted.txt:4]: See my previous comment but I dont think this is necessary

File: Week3/homework/ex1-Normalization/2-entitiesExtracted.txt [hunk 4]
```diff
@@ -0,0 +1,9 @@
...
+2- food_code: To avoid insert multiple data, we can extract this column into multiple columns (food_code1,food_code2,food_code3,.....).
+3- food_description: To avoid insert multiple data, we can extract this column into multiple columns (food_description1,food_description2,food_description3,.....).
+ or: To solve columns food_code,food_description, we can repeat rows for members which has multiple food_code and food_description and keep only 1 value in column.
+ or: We can use packages for choose the date form like: Carbon for Laravel and Dateformat for Node.js  
+4- dinner_date: To avoid inset data in different domains, we can extract this column into 3 columns (dinner_day hold integer data type,dinner_month hold integer data type,dinner_year hold integer data type[we can make it default = current year]).
```
Comment (mogaleaf) [inline Week3/homework/ex1-Normalization/2-entitiesExtracted.txt:9]: Or use a timestamp type ? This the db itself would force you to use same data type.

File: Week3/homework/ex1-Normalization/3-tables3NF.txt [hunk 5]
```diff
@@ -0,0 +1,16 @@
+3NF Rules (2NF rules + rule)
+    No transitive dependency (i.e. no column should depend on non-key column).
+
+To make the table in 3NF, we have to create 4 extra tables:
+    1- building table : has 2 columns: building_no(int primary key)     , building_name(varchar).
+    2- dinner table   : has 2 columns: dinner_id(int primary key)       , dinner_date(date).
+    3- venue table    : has 2 columns: venue_code(varchar primary key)  , venue_description(varchar).
+    4- food table     : has 2 columns: food_code(varchar primary key)   , food_description(varchar).
+
+And we still have member table which has 6 columns:
```
Comment (mogaleaf) [inline Week3/homework/ex1-Normalization/3-tables3NF.txt:10]: How do you link multiple food code for one member in one dinner in this configuration ?
Plus, the address is the member address, do you think it should be put along wit dinner ?
I think you still need one or two more tables.

File: Week3/homework/ex2-Transactions/transaction.js [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+const execQuery = util.promisify(connection.query.bind(connection));
+
+async function seedDatabase() {
+  connection.connect();
+
+  try {
+    // await execQuery('SET AUTOCOMMIT = 0;'); 
+    await execQuery('START TRANSACTION');
+  
+    await execQuery(`UPDATE account SET balance = balance - ${amountToTransfer} WHERE account_number = 101;`);
```
Comment (mogaleaf) [inline Week3/homework/ex2-Transactions/transaction.js:22]: What did you learn from exercice 3 ;) try to apply the rule here, really important specifically in money transaction :D

File: Week3/homework/ex2-Transactions/transaction.js [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    await execQuery(`UPDATE account SET balance = balance + ${amountToTransfer} WHERE account_number = 102;`);
+
+    await execQuery(`INSERT INTO account_changes(account_number,amount,changed_date,remark) VALUES(101,${amountToTransfer},now(),'Withdraw Money For Gift');`);
+    await execQuery(`INSERT INTO account_changes(account_number,amount,changed_date,remark) VALUES(102,${amountToTransfer},now(),'Received Money As Gift');`);
+
+    await execQuery("COMMIT");
```
Comment (mogaleaf) [inline Week3/homework/ex2-Transactions/transaction.js:28]: good

File: Week3/homework/ex2-Transactions/transaction.js [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+const amountToTransfer = 1000;
+const execQuery = util.promisify(connection.query.bind(connection));
+
+async function seedDatabase() {
+  connection.connect();
+
+  try {
+    // await execQuery('SET AUTOCOMMIT = 0;'); 
+    await execQuery('START TRANSACTION');
```
Comment (mogaleaf) [inline Week3/homework/ex2-Transactions/transaction.js:20]: good

File: Week3/homework/ex2-Transactions/transaction.js [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+    await execQuery(`UPDATE account SET balance = balance - ${amountToTransfer} WHERE account_number = 101;`);
+    await execQuery(`UPDATE account SET balance = balance + ${amountToTransfer} WHERE account_number = 102;`);
+
+    await execQuery(`INSERT INTO account_changes(account_number,amount,changed_date,remark) VALUES(101,${amountToTransfer},now(),'Withdraw Money For Gift');`);
+    await execQuery(`INSERT INTO account_changes(account_number,amount,changed_date,remark) VALUES(102,${amountToTransfer},now(),'Received Money As Gift');`);
+
+    await execQuery("COMMIT");
+  } catch (error) {
+    console.error(error);
+    await execQuery("ROLLBACK");
```
Comment (mogaleaf) [inline Week3/homework/ex2-Transactions/transaction.js:31]: good

File: Week3/homework/ex3-SQL Injection/1-valueNameAndCode.txt [hunk 10]
```diff
@@ -0,0 +1,14 @@
+1- Give an example of a value that can be passed as name and code that would take advantage of SQL-injection
```
Comment (mogaleaf) [inline Week3/homework/ex3-SQL Injection/1-valueNameAndCode.txt:1]: ok, and hacker could also drop schema for instance this way not only retrieve unexpected data.

File: Week3/homework/ex3-SQL Injection/2-rewriteTheFunction.js [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+        // And if I try to run multi statements (as an injection) it won't work! the query will be: SELECT Population FROM country WHERE Name = '\'s OR \'1\'=\'1' AND Code = '\'s OR \'1\'=\'1; show tables;
+
+        // const select_query = `SELECT Population FROM ${connection.escape(input_table_name).replace(/'/g,"")} WHERE Name = ${connection.escape(input_country_name)} AND Code = ${connection.escape(input_country_code)};`;
+       
+        // 3. Using a question mark syntax to do the escaping 
+        // This query will not print all countries populations, even if I write s' OR '1'='1 as country name or country code or both of them!!!
+        // So the query will be: SELECT Population FROM country WHERE Name = ? AND Code = ?;
+        // And if I try to run multi statements (as an injection) it won't work! the query will be: SELECT Population FROM country WHERE Name = ? AND Code = ?;
+
+        const select_query = `SELECT Population FROM ${connection.escape(input_table_name).replace(/'/g,"")} WHERE Name = ? AND Code = ?;`;
```
Comment (mogaleaf) [inline Week3/homework/ex3-SQL Injection/2-rewriteTheFunction.js:53]: I like you tried to escape the column name. But there is a better way to do it :
https://github.com/mysqljs/mysql#escaping-query-values
find the '??' marker.

File: Week3/homework/ex4-MongoDB CRUD/mongoDB.js [hunk 12]
```diff
@@ -0,0 +1,46 @@
+
+const MongoClient = require('mongodb').MongoClient;
```
Comment (mogaleaf) [inline Week3/homework/ex4-MongoDB CRUD/mongoDB.js:2]: ok

**Review by mogaleaf — COMMENTED** [review]
Hello,
Overall, it's a good homework.
You understand all the concepts of each exercises.
Just some small adjustment to make (see my comments) but nice job !

File: Week3/homework/ex3-SQL Injection/1-valueNameAndCode.txt (hunk 10 above)
Comment (KhaledAlsaleh) [reply Week3/homework/ex3-SQL Injection/1-valueNameAndCode.txt:1]: Yes sure, who can retrieve unexpected data, also can easily drop a schema.

File: Week3/homework/ex4-MongoDB CRUD/mongoDB.js (hunk 12 above)
Comment (KhaledAlsaleh) [reply Week3/homework/ex4-MongoDB CRUD/mongoDB.js:2]: Thanks for your time 💯.

Comment (KhaledAlsaleh) [reply Week3/homework/ex4-MongoDB CRUD/mongoDB.js:2]: Thanks a lot. ❤️

File: Week3/homework/ex3-SQL Injection/2-rewriteTheFunction.js (hunk 11 above)
Comment (KhaledAlsaleh) [reply Week3/homework/ex3-SQL Injection/2-rewriteTheFunction.js:53]: I just learned something new 👍🏻 👍🏻 It works with ?? marker!
I updated this file!

File: Week3/homework/ex2-Transactions/transaction.js (hunk 9 above)
Comment (KhaledAlsaleh) [reply Week3/homework/ex2-Transactions/transaction.js:31]: Thanks a lot! 👍🏻

File: Week3/homework/ex2-Transactions/transaction.js (hunk 8 above)
Comment (KhaledAlsaleh) [reply Week3/homework/ex2-Transactions/transaction.js:20]: 👍🏻 💯

File: Week3/homework/ex2-Transactions/transaction.js (hunk 7 above)
Comment (KhaledAlsaleh) [reply Week3/homework/ex2-Transactions/transaction.js:28]: Thank you 👍🏻

File: Week3/homework/ex2-Transactions/transaction.js (hunk 6 above)
Comment (KhaledAlsaleh) [reply Week3/homework/ex2-Transactions/transaction.js:22]: You are right! I have thought about that, but here I didn't expect any input from a user, so I didn't implement the secure code!
Anyway, I updated the file! 😄 🚀

File: Week3/homework/ex1-Normalization/1-violate1NF.txt (hunk 2 above)
Comment (KhaledAlsaleh) [reply Week3/homework/ex1-Normalization/1-violate1NF.txt:7]: True!

File: Week3/homework/ex1-Normalization/2-entitiesExtracted.txt (hunk 4 above)
Comment (KhaledAlsaleh) [reply Week3/homework/ex1-Normalization/2-entitiesExtracted.txt:9]: Yes, I can use timestamp type It's really easier!

File: Week3/homework/ex1-Normalization/1-violate1NF.txt (hunk 1 above)
Comment (KhaledAlsaleh) [reply Week3/homework/ex1-Normalization/1-violate1NF.txt:6]: The member_address  column violates Rule1+2 [I also thought it's not necessary 😄 ].
But I follow the reading material and videos: https://www.youtube.com/watch?v=G6nX_0lY75Y&list=PLVYDhqbgYpYVq-rHFB1x2-P7esnWotCB7&index=11&t=995s (min 17:30)

File: Week3/homework/ex1-Normalization/2-entitiesExtracted.txt (hunk 3 above)
Comment (KhaledAlsaleh) [reply Week3/homework/ex1-Normalization/2-entitiesExtracted.txt:4]: Done! Thanks a lot for your time. 💯

File: Week3/homework/ex1-Normalization/3-tables3NF.txt (hunk 5 above)
Comment (KhaledAlsaleh) [reply Week3/homework/ex1-Normalization/3-tables3NF.txt:10]: If the member has multiple foods in one dinner, then we need a table to link all food codes with the member. And we could have an extra table for adding member address with dinner details.
Sure we can have more tables...

Thanks a lot. 👍🏻

**Comment by KhaledAlsaleh** [comment]
> Hello,
> Overall, it's a good homework.
> You understand all the concepts of each exercises.
> Just some small adjustment to make (see my comments) but nice job !

Hello @mogaleaf,

Thanks a lot for the feedback, it's really useful and I have learned a new thing!
I updated the files, hope it's good now.

File: Week3/homework/ex1-Normalization/1-violate1NF.txt (hunk 1 above)
Comment (KhaledAlsaleh) [reply Week3/homework/ex1-Normalization/1-violate1NF.txt:6]: I updated the file, Considering that the member_address  is a string.
Thanks.

**Pull request closed without merging**